tests/logs.test.ts

```diff
@@ -0,0 +1,371 @@
+    const logSpy = jest.spyOn(console, "log");
+    const errorSpy = jest.spyOn(console, "error");
+    const debugSpy = jest.spyOn(console, "debug");
+    handler = new RPCHandler({ ...mods, proxySettings: { ...mods.proxySettings, strictLogs: false, logTier: "info" } });
+    provider = await handler.getFastestRpcProvider();
+
+    const response = await provider.send("eth_blockNumber", []);
+    expect(response).toBeDefined();
+    expect(parseInt(response)).toBeGreaterThan(0);
+
+    const cleanLogStrings = cleanSpyLogs(logSpy);
+    expect(cleanLogStrings).toEqual(expect.arrayContaining([cleanLogString(INITIALIZED)]));
+
+    const cleanErrorStrings = cleanSpyLogs(errorSpy);
+    expect(cleanErrorStrings).toEqual([]);
+
+    const cleanDebugStrings = cleanSpyLogs(debugSpy);
+    expect(cleanDebugStrings).toEqual(expect.arrayContaining([cleanLogString(BLOCK_NUMBER_CALL), cleanLogString(BLOCK_NUMBER_CALL)]));
+  });
+
+  it("should have multiple logs when retrying failed requests", async () => {
+    const logSpy = jest.spyOn(console, "log");
+    const errorSpy = jest.spyOn(console, "error");
+    const debugSpy = jest.spyOn(console, "debug");
+    const warnSpy = jest.spyOn(console, "warn");
+
+    const badRpcs = ["http://127.0.0.1:8546", "http://127.0.0.1:8544", "http://127.0.0.1:8544", "http://127.0.0.1:8544", "http://127.0.0.1:8545"];
+
+    handler = new RPCHandler({
+      ...mods,
+      networkRpcs: badRpcs,
+      proxySettings: { ...mods.proxySettings, logTier: "verbose", strictLogs: false, retryCount: 5, retryDelay: 20 },
+    });
+    provider = await handler.getFastestRpcProvider();
+
+    try {
+      // this call will always fail and bad rpcs are filtered out
+      // so it retries 5 times then fails
+      const response = await provider.send("eth_call", [null, null, null, "latest"]);
+      expect(response).toBeDefined();
+      expect(response).toBe("0x" + "00".repeat(32));
+    } catch (error) {
+      expect(error).toBeDefined();
+      expect(error).toBeInstanceOf(Error);
+    }
+
+    let cleanErrorStrings = cleanSpyLogs(errorSpy);
+    const filteredStrings = cleanErrorStrings.filter((str) => str.includes(cleanLogString(NULL_ARG_TX_CALL)));
+    expect(filteredStrings.length).toBeGreaterThanOrEqual(2);
+
+    let cleanDebugStrings = cleanSpyLogs(debugSpy);
+    expect(cleanDebugStrings).toEqual(
+      expect.arrayContaining([
+        cleanLogString(DEBUG_CONNECT_TO),
+        cleanLogString(DEBUG_RETRY),
+        cleanLogString(DEBUG_RETRY_IN_20),
+        cleanLogString(`${DEBUG_CALL_NUMBER}1`),
+        cleanLogString(`${DEBUG_CALL_NUMBER}2`),
+        cleanLogString(`${DEBUG_CALL_NUMBER}3`),
+        cleanLogString(`${DEBUG_CALL_NUMBER}4`),
+      ])
+    );
+
+    debugSpy.mockClear();
+    errorSpy.mockClear();
+
+    const res = await provider.send("eth_call", [nonceBitmapData, "latest"]);
+    expect(res).toBeDefined();
+    expect(res).toBe("0x" + "00".repeat(32));
```

Comment:
ha! catch it!

Reply:
i mean the last one was 32 bytes we could have done an assertion, this seems like a innovative way 😄